File: ios/HybridPharmaScanner.swift
import NitroModules
import AVFoundation
import UIKit

class HybridPharmaScanner: HybridPharmaScannerSpec {
  private let cameraManager = CameraManager.shared
  private let documentScannerManager = DocumentScannerManager()

  func ping() throws -> String {
    return "pong"
  }

  func getVersion() throws -> String {
    return "0.0.1"
  }

  func startCamera() throws -> Void {
    Task {
      let granted = await cameraManager.requestCameraPermission()
      guard granted else {
        print("[PharmaScanner] Camera permission denied")
        return
      }
      try cameraManager.startSession()
    }
  }

  func stopCamera() throws -> Void {
    cameraManager.stopSession()
  }

  func capturePhoto() throws -> Promise<CapturedImage> {
    return Promise.async {
      let (data, width, height) = try await self.cameraManager.capturePhoto()

      // Save JPEG to temp directory
      let fileName = UUID().uuidString + ".jpg"
      let tempDir = FileManager.default.temporaryDirectory
      let fileURL = tempDir.appendingPathComponent(fileName)
      try data.write(to: fileURL)

      return CapturedImage(
        uri: fileURL.absoluteString,
        width: Double(width),
        height: Double(height),
        base64: nil
      )
    }
  }

  func setFlash(mode: FlashMode) throws -> Void {
    let avFlashMode: AVCaptureDevice.FlashMode
    switch mode {
    case .auto:
      avFlashMode = .auto
    case .on:
      avFlashMode = .on
    case .off:
      avFlashMode = .off
    }
    cameraManager.setFlashMode(avFlashMode)
  }

  func setZoom(factor: Double) throws -> Void {
    cameraManager.setZoom(factor: factor)
  }

  func detectDocument(imageUri: String) throws -> Promise<DocumentDetection> {
    return Promise.async {
      let detector = DocumentDetector()
      return try await detector.detectDocument(imageUri: imageUri)
    }
  }

  func cropAndCorrect(imageUri: String, corners: Corners) throws -> Promise<CapturedImage> {
    return Promise.async {
      let (data, width, height) = try ImageProcessor.cropAndCorrect(imageUri: imageUri, corners: corners)

      let fileName = UUID().uuidString + ".jpg"
      let tempDir = FileManager.default.temporaryDirectory
      let fileURL = tempDir.appendingPathComponent(fileName)
      try data.write(to: fileURL)

      return CapturedImage(
        uri: fileURL.absoluteString,
        width: Double(width),
        height: Double(height),
        base64: nil
      )
    }
  }

  func setOnDocumentDetected(callback: @escaping (_ detection: DocumentDetection) -> Void) throws -> Void {
    cameraManager.setOnDocumentDetected(callback)
  }

  func scanDocument() throws -> Promise<[CapturedImage]> {
    return Promise.async {
      return try await self.documentScannerManager.scanDocument()
    }
  }

  func scanBarcodes(options: BarcodeScanOptions) throws -> Promise<[BarcodeResult]> {
    return Promise.async {
      let scanner = BarcodeScanner()
      return try await scanner.scanBarcodes(imageUri: options.imageUri, formats: options.formats)
    }
  }

  func startContinuousScan(formats: [BarcodeFormat], onDetected: @escaping (_ codes: [BarcodeResult]) -> Void) throws -> Void {
    cameraManager.onBarcodesDetectedCallback = onDetected
    cameraManager.startContinuousScan(formats: formats)
  }

  func stopContinuousScan() throws -> Void {
    cameraManager.stopContinuousScan()
  }

  func recognizeText(imageUri: String) throws -> Promise<OcrResult> {
    return Promise.async {
      let processor = OcrProcessor()
      return try await processor.recognizeText(imageUri: imageUri)
    }
  }

  func recognizeDocument(imageUri: String) throws -> Promise<OcrResult> {
    return Promise.async {
      let processor = DocumentOcrProcessor()
      return try await processor.recognizeDocument(imageUri: imageUri)
    }
  }

  func setOnTextRecognized(callback: @escaping (_ result: OcrResult) -> Void) throws -> Void {
    cameraManager.onTextRecognizedCallback = callback
  }

  func scanInvoice(imageUri: String) throws -> Promise<InvoiceResult> {
    return Promise.async {
      let processor = OcrProcessor()
      let ocrResult = try await processor.recognizeText(imageUri: imageUri)
      let parser = InvoiceParser()
      return parser.parse(ocrResult: ocrResult)
    }
  }

  func configure(apiKey: String, baseUrl: String) throws -> Void {
    DocumentExtractor.shared.configure(apiKey: apiKey, baseUrl: baseUrl)
  }

  func recognizeStructuredDocument(imageUri: String) throws -> Promise<StructuredDocumentResult> {
    return Promise.async {
      let processor = DocumentOcrProcessor()
      return try await processor.recognizeStructuredDocument(imageUri: imageUri)
    }
  }

  func extractDocument(imageUri: String, options: ExtractionOptions) throws -> Promise<DocumentExtractionResult> {
    return Promise.async {
      return try await DocumentExtractor.shared.extract(
        imageUri: imageUri,
        documentType: String(options.documentType),
        language: String(options.language),
        customPrompt: options.customPrompt.flatMap { String($0) },
        forceOffline: options.forceOffline ?? false,
        scanOcr: options.scanOcr ?? false
      )
    }
  }
}

File: pharmascanner/src/main/java/com/margelo/nitro/PharmaScanner/HybridPharmaScanner.kt
package com.margelo.nitro.PharmaScanner

import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.lifecycle.LifecycleOwner
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class HybridPharmaScanner : HybridPharmaScannerSpec() {

    companion object {
        private const val TAG = "HybridPharmaScanner"
        private val mainHandler = Handler(Looper.getMainLooper())
    }

    override fun ping(): String = "pong"
    override fun getVersion(): String = "0.0.1"

    override fun startCamera(): Unit {
        mainHandler.post {
            val activity = ActivityProvider.currentActivity ?: run {
                Log.w(TAG, "startCamera: no current activity")
                return@post
            }

            val componentActivity = activity as? ComponentActivity ?: run {
                Log.w(TAG, "startCamera: activity is not a ComponentActivity")
                return@post
            }

            val lifecycleOwner = activity as? LifecycleOwner ?: run {
                Log.w(TAG, "startCamera: activity is not a LifecycleOwner")
                return@post
            }

            PermissionHelper.requestCameraPermission(componentActivity) { granted ->
                if (granted) {
                    CameraManager.startSession(activity, lifecycleOwner)
                } else {
                    Log.w(TAG, "Camera permission denied")
                }
            }
        }
    }

    override fun stopCamera(): Unit {
        CameraManager.stopSession()
    }

    override fun capturePhoto(): Promise<CapturedImage> {
        return Promise.async {
            val activity = ActivityProvider.currentActivity
                ?: throw IllegalStateException("Activity not available")

            val (uri, width, height) = CameraManager.capturePhoto(activity)

            CapturedImage(
                uri = uri,
                width = width.toDouble(),
                height = height.toDouble(),
                base64 = null
            )
        }
    }

    override fun setFlash(mode: FlashMode): Unit {
        CameraManager.setFlashMode(mode)
    }

    override fun setZoom(factor: Double): Unit {
        CameraManager.setZoom(factor)
    }

    override fun detectDocument(imageUri: String): Promise<DocumentDetection> {
        return Promise.async {
            throw UnsupportedOperationException("Use scanDocument() on Android")
        }
    }

    override fun cropAndCorrect(imageUri: String, corners: Corners): Promise<CapturedImage> {
        return Promise.async {
            throw UnsupportedOperationException("Use scanDocument() on Android")
        }
    }

    override fun setOnDocumentDetected(callback: (detection: DocumentDetection) -> Unit): Unit {
        Log.d(TAG, "setOnDocumentDetected is a no-op on Android. Use scanDocument() instead.")
    }

    override fun scanDocument(): Promise<Array<CapturedImage>> {
        return Promise.async {
            val result = DocumentScannerManager.scanDocument()
            result.toTypedArray()
        }
    }

    override fun scanBarcodes(options: BarcodeScanOptions): Promise<Array<BarcodeResult>> {
        return Promise.async {
            BarcodeScannerManager.scanBarcodes(options)
        }
    }

    override fun startContinuousScan(formats: Array<BarcodeFormat>, onDetected: (codes: Array<BarcodeResult>) -> Unit): Unit {
        BarcodeScannerManager.onBarcodesDetectedCallback = onDetected
        BarcodeScannerManager.activeFormats = formats
        CameraManager.startContinuousScan(formats)
    }

    override fun stopContinuousScan(): Unit {
        CameraManager.stopContinuousScan()
    }

    override fun recognizeText(imageUri: String): Promise<OcrResult> {
        return Promise.async {
            OcrManager.recognizeText(imageUri)
        }
    }

    override fun setOnTextRecognized(callback: (result: OcrResult) -> Unit): Unit {
        OcrManager.onTextRecognizedCallback = callback
        Log.d(TAG, "setOnTextRecognized: registered callback. Use recognizeText() with image URI on Android.")
    }

    override fun recognizeDocument(imageUri: String): Promise<OcrResult> {
        return Promise.async {
            OcrManager.recognizeText(imageUri)
        }
    }

    override fun configure(apiKey: String, baseUrl: String): Unit {
        DocumentExtractor.configure(apiKey, baseUrl)
    }

    override fun extractDocument(imageUri: String, options: ExtractionOptions): Promise<DocumentExtractionResult> {
        Log.d(TAG, "extractDocument called: mode=${options.customPrompt}, type=${options.documentType}")
        return Promise.async {
            // Run entire extraction off the main thread to prevent ANR
            withContext(Dispatchers.Default) {
                Log.d(TAG, "extractDocument coroutine started on ${Thread.currentThread().name}")
                DocumentExtractor.extract(
                    imageUri = imageUri,
                    documentType = options.documentType,
                    language = options.language,
                    customPrompt = options.customPrompt,
                    forceOffline = options.forceOffline ?: false
                )
            }
        }
    }

    override fun isLocalLlmModelReady(): Boolean {
        return try {
            LlamaCppManager.isModelDownloaded
        } catch (e: Exception) {
            Log.w(TAG, "isLocalLlmModelReady check failed", e)
            false
        }
    }

    override fun downloadLocalLlmModel(onProgress: (progress: Double) -> Unit): Promise<Unit> {
        return Promise.async {
            LlamaCppManager.downloadModel { progress ->
                onProgress(progress)
            }
        }
    }

    override fun unloadLocalLlmModel(): Unit {
        LlamaCppManager.unloadModel()
    }
}
